src/GitHub.Api/Git/Repository.cs

```diff
@@ -309,273 +207,97 @@ private void CacheContainer_OnCacheInvalidated(CacheType cacheType)
-        }
-
-        private void HandleBranchCacheUpdatedEvent(CacheUpdateEvent cacheUpdateEvent)
-        {
-            Logger.Trace("BranchCache Updated {0}", cacheUpdateEvent.UpdatedTimeString);
-            LocalBranchListChanged?.Invoke(cacheUpdateEvent);
-            RemoteBranchListChanged?.Invoke(cacheUpdateEvent);
-            LocalAndRemoteBranchListChanged?.Invoke(cacheUpdateEvent);
-        }
-
         private void RepositoryManagerOnCurrentBranchUpdated(ConfigBranch? branch, ConfigRemote? remote)
         {
-            new ActionTask(TaskManager.Instance.Token, () => {
-                if (!Nullable.Equals(CurrentConfigBranch, branch))
-                {
-                    var currentBranch = branch != null ? (GitBranch?)GetLocalGitBranch(branch.Value) : null;
-
-                    CurrentConfigBranch = branch;
-                    CurrentBranch = currentBranch;
-                    UpdateLocalBranches();
-                }
-
-                if (!Nullable.Equals(CurrentConfigRemote, remote))
-                {
-                    CurrentConfigRemote = remote;
-                    CurrentRemote = remote.HasValue ? (GitRemote?)GetGitRemote(remote.Value) : null;
-                    ClearRepositoryInfo();
-                }
-            }) { Affinity = TaskAffinity.UI }.Start();
+            taskManager.RunInUI(() =>
+            {
+                var data = new RepositoryInfoCacheData();
+                data.CurrentConfigBranch = branch;
+                data.CurrentGitBranch = branch.HasValue ? (GitBranch?)GetLocalGitBranch(branch.Value.name, branch.Value) : null;
+                data.CurrentConfigRemote = remote;
+                data.CurrentGitRemote = remote.HasValue ? (GitRemote?)GetGitRemote(remote.Value) : null;
+                name = null;
+                cloneUrl = null;
+                cacheContainer.RepositoryInfoCache.UpdateData(data);
+                var n = Name; // force refresh of the Name and CloneUrl property
+                // update active state in local branches
+                cacheContainer.BranchCache.LocalBranches = LocalConfigBranches;
+                // update tracking state in remote branches
+                cacheContainer.BranchCache.RemoteBranches = RemoteConfigBranches;
+            });
         }
 
         private void RepositoryManagerOnGitStatusUpdated(GitStatus gitStatus)
         {
-            new ActionTask(TaskManager.Instance.Token, () => {
-                CurrentChanges = gitStatus.Entries;
-                CurrentAhead = gitStatus.Ahead;
-                CurrentBehind = gitStatus.Behind;
-            }) { Affinity = TaskAffinity.UI }.Start();
+            taskManager.RunInUI(() =>
+            {
+                cacheContainer.GitStatusEntriesCache.Entries = gitStatus.Entries;
+                cacheContainer.GitTrackingStatusCache.Ahead = gitStatus.Ahead;
+                cacheContainer.GitTrackingStatusCache.Behind = gitStatus.Behind;
```

Comment:
Yes, I see how we would want a single structure to handle the updates here and below.